src/dune_pkg/opam_solver.ml

```diff
@@ -0,0 +1,190 @@
+    | Right r -> R.candidates r name |> convert_rejections ~f:Either.right
+
+  let user_restrictions = function
+    | Left l -> L.user_restrictions l
+    | Right r -> R.user_restrictions r
+
+  let filter_deps = function
+    | Left l -> L.filter_deps l
+    | Right r -> R.filter_deps r
+end
+
+(* Helper functor which adds a set of local packages to a [CONTEXT] *)
+module Context_with_local_packages (Base_context : CONTEXT) : sig
+  include CONTEXT
+
+  val create :
+       base_context:Base_context.t
+    -> local_packages:OpamFile.OPAM.t OpamPackage.Name.Map.t
+    -> t
+
+  val local_packages : t -> OpamFile.OPAM.t OpamPackage.Name.Map.t
+end = struct
+  let local_package_default_version = OpamPackage.Version.of_string "LOCAL"
+
+  type t =
+    { base_context : Base_context.t
+    ; local_packages : OpamFile.OPAM.t OpamPackage.Name.Map.t
+    }
+
+  let create ~base_context ~local_packages = { base_context; local_packages }
+
+  let local_packages { local_packages; _ } = local_packages
+
+  type rejection = Base_context.rejection
+
+  let pp_rejection = Base_context.pp_rejection
+
+  let candidates t name =
+    match OpamPackage.Name.Map.find_opt name t.local_packages with
+    | None -> Base_context.candidates t.base_context name
+    | Some opam_file ->
+      let version =
+        Option.value opam_file.version ~default:local_package_default_version
+      in
+      [ (version, Ok opam_file) ]
+
+  let user_restrictions t = Base_context.user_restrictions t.base_context
+
+  let filter_deps t = Base_context.filter_deps t.base_context
+end
+
+(* Helper function for implementing [CONTEXT.filter_deps] where local package
+   dependencies include test and doc dependencies *)
+let filter_deps_env ~local_packages ~env pkg filtered_formula =
+  let pkg_is_local =
+    OpamPackage.Name.Map.mem (OpamPackage.name pkg) local_packages
+  in
+  filtered_formula
+  |> OpamFilter.partial_filter_formula (env pkg)
+  |> OpamFilter.filter_deps ~build:true ~post:true ~test:pkg_is_local
+       ~doc:pkg_is_local ~dev:pkg_is_local ~default:false
+
+module Dir_context_for_dune = struct
+  module Dir_context = Opam_0install.Dir_context
+  module Base_context = Context_with_local_packages (Dir_context)
+
+  (* Wrapper of [Opam_0install.Dir_context] which exposes its environment *)
+  type t =
+    { base_context : Base_context.t
+    ; env : string -> OpamVariable.variable_contents option
+    }
+
+  let create packages_dir_path ~env ~local_packages ~prefer_oldest =
+    let base_context =
+      Base_context.create
+        ~base_context:
+          (Dir_context.create ~prefer_oldest
+             ~constraints:OpamPackage.Name.Map.empty ~env packages_dir_path)
+        ~local_packages
+    in
+    { base_context; env }
+
+  (* Override [filter_deps] so that local packages get test and doc dependencies *)
+  let filter_deps t =
+    let local_packages = Base_context.local_packages t.base_context in
+    filter_deps_env ~local_packages ~env:(fun pkg v ->
```

Comment:
Isn't this the env function from the solver? Maybe it can be reused?

Reply:
Actually yes, I've since dug a bit deeper into opam filters in another change I'm working on and I think we can simplify this by resolving the `test` and `doc` variables to true before calling the base context's `filter_deps` on the result. Will update.